In [1]:
import pandas as pd
df=pd.read_csv('/Users/lunlun/Desktop/BDSE_18_Project/data/85年到110年8個欄位.csv',low_memory=False)
df

交易日期 作物代號  市場代號     上價     中價    下價    平均價     交易量
0         1996-01-01   71   109   13.0    7.0   3.0    9.0  7225.0
1         1996-01-01   45   104  200.0  127.3  80.0  128.0   147.0
2         1996-01-01   71   241    9.0    7.0   5.0    8.6  1580.0
3         1996-01-01   72   600   35.0   30.0  20.0   30.0  1920.0
4         1996-01-01   A1   830   27.0   17.9  10.0   17.9  4010.0
...              ...  ...   ...    ...    ...   ...    ...     ...
20152741  2021-03-01  SJ2   950   72.3   68.0  46.9   64.7  1048.0
20152742  2021-03-01  SO0   950    9.0    8.6   7.6    8.5   312.0
20152743  2021-03-01  SO1   950   18.0   15.8  13.5   15.8   285.0
20152744  2021-03-01  SP1   950   37.1   26.1  16.0   26.3   365.0
20152745  2021-03-01  SW1   950   11.9    9.2   7.6    9.4   842.0

[20152746 rows x 8 columns]

N04 蔬菜          N05 水果        N06 花卉
---------------------------------------------------
109 台北一        109 台北一      105 台北(市場) 這邊都是花的部分
104 台北二        104 台北二      
220 板橋(區)      220 板橋(區)
241 三重(區)      241 三重(區)    514 彰化(市場)
260 宜蘭(市)      260 宜蘭(市)
338 桃園(桃農)    338 桃園(桃農)  700 台南(市場)
400 台中(市)      400 台中(市)    400 台中(市場)
420 豐原(區)      420 豐原(區)
                 423 東勢(鎮)(台中)
540 南投(市)      540 南投(市)
512 永靖(鄉)彰化  600 嘉義(市)
514 溪湖(鎮)雲林 
648 西螺(鎮) 
800 高雄(市)      800 高雄(市)    800 高雄(市場)
830 鳳山(區)      830 鳳山(區)
900 屏東(市)    
930 台東(市)      930 台東(市)
950 花蓮(市)

424 -> 郵遞區號為 台中市 和平區

In [11]:
ma=df['作物代號']=='FF1'
df1=df[ma]
# df[df['作物代號']=='LF2']

In [12]:
df1

交易日期 作物代號  市場代號    上價    中價    下價   平均價      交易量
597       1996-01-01  FF1   104  30.0  20.0  15.0  23.1   1160.0
598       1996-01-01  FF1   109  30.0  20.0  10.0  21.5   2341.0
782       1996-01-01  FF1   648  23.0  20.0  15.0  19.6    312.0
783       1996-01-01  FF1   800  25.0  20.0  15.0  19.6    542.0
1090      1996-01-01  FF1   400  22.5  17.5  12.5  17.5     40.0
...              ...  ...   ...   ...   ...   ...   ...      ...
20150822  2021-03-02  FF1   800  35.0  30.9  23.5  30.2   3468.0
20151146  2021-03-02  FF1   830  30.1  19.0  10.0  19.4   1750.0
20151233  2021-03-02  FF1   900  25.0  16.7  10.0  16.7  11591.0
20151366  2021-03-02  FF1   930  35.0  17.0  10.0  17.0    640.0
20152271  2021-03-01  FF1   648  49.4  38.9  19.5  37.3  20666.0

[91212 rows x 8 columns]

選擇
<!--  -->
白蘿蔔 SA3 千仔
<!--  -->
絲瓜 ff1 91212

In [13]:
ma1=df1['市場代號']==109
ma2=df1['市場代號']==104
df2=df1[ma1|ma2]

In [14]:
ma3=df2['交易日期']>'1999-12-31'
df3=df2[ma3]

In [15]:
# di1={104:'Taipei',109:'Taipei'}
df3['市場代號']=df3['市場代號'].replace(109,'Taipei')
df3['市場代號']=df3['市場代號'].replace(104,'Taipei')

<ipython-input-15-34b3119a4407>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['市場代號']=df3['市場代號'].replace(109,'Taipei')
<ipython-input-15-34b3119a4407>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['市場代號']=df3['市場代號'].replace(104,'Taipei')


In [16]:
df3['市場代號'].unique()

array(['Taipei'], dtype=object)

In [17]:
df3

交易日期 作物代號    市場代號    上價    中價    下價   平均價      交易量
2226630   2000-01-02  FF1  Taipei  48.0  30.0  12.0  27.5   6106.0
2227660   2000-01-02  FF1  Taipei  35.4  26.5  14.4  25.9   1242.0
2228209   2000-01-03  FF1  Taipei  28.8  19.4  12.9  20.0    793.0
2229463   2000-01-03  FF1  Taipei  40.0  25.0  10.0  26.6   3459.0
2230326   2000-01-04  FF1  Taipei  35.0  21.0   6.0  18.0   4931.0
...              ...  ...     ...   ...   ...   ...   ...      ...
20142476  2021-03-04  FF1  Taipei  34.0  21.7  11.4  22.1  14699.0
20145079  2021-03-03  FF1  Taipei  27.8  19.0  10.7  19.1   7194.0
20145666  2021-03-03  FF1  Taipei  32.4  18.2  10.4  19.5  16756.0
20148307  2021-03-02  FF1  Taipei  40.4  26.2  14.6  26.7   8004.0
20148880  2021-03-02  FF1  Taipei  32.3  19.7  10.8  20.4  37548.0

[12838 rows x 8 columns]

In [18]:
test=df3.iloc[:,[0,4]]
test=test.groupby('交易日期').apply(lambda x:x['中價'].mean())
test
test=pd.DataFrame(test)
test

0
交易日期             
2000-01-02  28.25
2000-01-03  22.20
2000-01-04  23.40
2000-01-05  19.65
2000-01-06  20.75
...           ...
2021-03-26  32.95
2021-03-27  34.35
2021-03-28  30.30
2021-03-30  25.35
2021-03-31  22.00

[6419 rows x 1 columns]

In [19]:
test1=df3.iloc[:,[0,7]]
# test1
test1=test1.groupby('交易日期').apply(lambda x:x['交易量'].sum())
test1
test1=pd.DataFrame(test1)
test1

0
交易日期               
2000-01-02   7348.0
2000-01-03   4252.0
2000-01-04   5486.0
2000-01-05   4841.0
2000-01-06   5049.0
...             ...
2021-03-26  18932.0
2021-03-27  23591.0
2021-03-28  31770.0
2021-03-30  61186.0
2021-03-31  39858.0

[6419 rows x 1 columns]

In [20]:
df4=df3.iloc[:,[2,0]]
df4

市場代號        交易日期
2226630   Taipei  2000-01-02
2227660   Taipei  2000-01-02
2228209   Taipei  2000-01-03
2229463   Taipei  2000-01-03
2230326   Taipei  2000-01-04
...          ...         ...
20142476  Taipei  2021-03-04
20145079  Taipei  2021-03-03
20145666  Taipei  2021-03-03
20148307  Taipei  2021-03-02
20148880  Taipei  2021-03-02

[12838 rows x 2 columns]

In [21]:
df5=df4.drop_duplicates()

In [22]:
df5

市場代號        交易日期
2226630   Taipei  2000-01-02
2228209   Taipei  2000-01-03
2230326   Taipei  2000-01-04
2232188   Taipei  2000-01-05
2234298   Taipei  2000-01-06
...          ...         ...
20135806  Taipei  2021-03-06
20138694  Taipei  2021-03-05
20141913  Taipei  2021-03-04
20145079  Taipei  2021-03-03
20148307  Taipei  2021-03-02

[6419 rows x 2 columns]

In [23]:
df6=pd.merge(df5,test1,on=['交易日期'])
# last_df=pd.merge(df,veg,on=['ObservDay_time'],how='outer')

In [24]:
df6

市場代號        交易日期        0
0     Taipei  2000-01-02   7348.0
1     Taipei  2000-01-03   4252.0
2     Taipei  2000-01-04   5486.0
3     Taipei  2000-01-05   4841.0
4     Taipei  2000-01-06   5049.0
...      ...         ...      ...
6414  Taipei  2021-03-06  19118.0
6415  Taipei  2021-03-05  20015.0
6416  Taipei  2021-03-04  20803.0
6417  Taipei  2021-03-03  23950.0
6418  Taipei  2021-03-02  45552.0

[6419 rows x 3 columns]

In [25]:
df7=pd.merge(df6,test,on=['交易日期'])

In [26]:
df7

市場代號        交易日期      0_x    0_y
0     Taipei  2000-01-02   7348.0  28.25
1     Taipei  2000-01-03   4252.0  22.20
2     Taipei  2000-01-04   5486.0  23.40
3     Taipei  2000-01-05   4841.0  19.65
4     Taipei  2000-01-06   5049.0  20.75
...      ...         ...      ...    ...
6414  Taipei  2021-03-06  19118.0  27.80
6415  Taipei  2021-03-05  20015.0  22.90
6416  Taipei  2021-03-04  20803.0  21.30
6417  Taipei  2021-03-03  23950.0  18.60
6418  Taipei  2021-03-02  45552.0  22.95

[6419 rows x 4 columns]

In [27]:
df8 = df7.sort_values(by = '交易日期')
df8.reset_index(inplace=True,drop=True)

In [28]:
df8

市場代號        交易日期      0_x    0_y
0     Taipei  2000-01-02   7348.0  28.25
1     Taipei  2000-01-03   4252.0  22.20
2     Taipei  2000-01-04   5486.0  23.40
3     Taipei  2000-01-05   4841.0  19.65
4     Taipei  2000-01-06   5049.0  20.75
...      ...         ...      ...    ...
6414  Taipei  2021-03-26  18932.0  32.95
6415  Taipei  2021-03-27  23591.0  34.35
6416  Taipei  2021-03-28  31770.0  30.30
6417  Taipei  2021-03-30  61186.0  25.35
6418  Taipei  2021-03-31  39858.0  22.00

[6419 rows x 4 columns]

In [29]:
df8.columns=['Market','ObservDay_time','TradingVolume','MediumAvgPrice']
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice
0     Taipei     2000-01-02         7348.0           28.25
1     Taipei     2000-01-03         4252.0           22.20
2     Taipei     2000-01-04         5486.0           23.40
3     Taipei     2000-01-05         4841.0           19.65
4     Taipei     2000-01-06         5049.0           20.75
...      ...            ...            ...             ...
6414  Taipei     2021-03-26        18932.0           32.95
6415  Taipei     2021-03-27        23591.0           34.35
6416  Taipei     2021-03-28        31770.0           30.30
6417  Taipei     2021-03-30        61186.0           25.35
6418  Taipei     2021-03-31        39858.0           22.00

[6419 rows x 4 columns]

In [30]:
import numpy as np
df8['T+1']=np.nan
df8['T+3']=np.nan
df8['T+5']=np.nan
df8['T+7']=np.nan
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice  T+1  T+3  T+5  T+7
0     Taipei     2000-01-02         7348.0           28.25  NaN  NaN  NaN  NaN
1     Taipei     2000-01-03         4252.0           22.20  NaN  NaN  NaN  NaN
2     Taipei     2000-01-04         5486.0           23.40  NaN  NaN  NaN  NaN
3     Taipei     2000-01-05         4841.0           19.65  NaN  NaN  NaN  NaN
4     Taipei     2000-01-06         5049.0           20.75  NaN  NaN  NaN  NaN
...      ...            ...            ...             ...  ...  ...  ...  ...
6414  Taipei     2021-03-26        18932.0           32.95  NaN  NaN  NaN  NaN
6415  Taipei     2021-03-27        23591.0           34.35  NaN  NaN  NaN  NaN
6416  Taipei     2021-03-28        31770.0           30.30  NaN  NaN  NaN  NaN
6417  Taipei     2021-03-30        61186.0           25.35  NaN  NaN  NaN  NaN
6418  Taipei     2021-03-31        39858.0           22.00  NaN  NaN  NaN  NaN

[6419 rows x 8 columns]

In [31]:
df8['ObservDay_time']=pd.to_datetime(df8['ObservDay_time'])

In [32]:
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice  T+1  T+3  T+5  T+7
0     Taipei     2000-01-02         7348.0           28.25  NaN  NaN  NaN  NaN
1     Taipei     2000-01-03         4252.0           22.20  NaN  NaN  NaN  NaN
2     Taipei     2000-01-04         5486.0           23.40  NaN  NaN  NaN  NaN
3     Taipei     2000-01-05         4841.0           19.65  NaN  NaN  NaN  NaN
4     Taipei     2000-01-06         5049.0           20.75  NaN  NaN  NaN  NaN
...      ...            ...            ...             ...  ...  ...  ...  ...
6414  Taipei     2021-03-26        18932.0           32.95  NaN  NaN  NaN  NaN
6415  Taipei     2021-03-27        23591.0           34.35  NaN  NaN  NaN  NaN
6416  Taipei     2021-03-28        31770.0           30.30  NaN  NaN  NaN  NaN
6417  Taipei     2021-03-30        61186.0           25.35  NaN  NaN  NaN  NaN
6418  Taipei     2021-03-31        39858.0           22.00  NaN  NaN  NaN  NaN

[6419 rows x 8 columns]

In [33]:
df8['T+1']=((df8['MediumAvgPrice'].shift(-1)-df8['MediumAvgPrice'])/df8['MediumAvgPrice']).apply(lambda x:format(x,'.3f'))
df8['T+3']=((df8['MediumAvgPrice'].shift(-3)-df8['MediumAvgPrice'])/df8['MediumAvgPrice']).apply(lambda x:format(x,'.3f'))
df8['T+5']=((df8['MediumAvgPrice'].shift(-5)-df8['MediumAvgPrice'])/df8['MediumAvgPrice']).apply(lambda x:format(x,'.3f'))
df8['T+7']=((df8['MediumAvgPrice'].shift(-7)-df8['MediumAvgPrice'])/df8['MediumAvgPrice']).apply(lambda x:format(x,'.3f'))

In [34]:
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice     T+1     T+3  \
0     Taipei     2000-01-02         7348.0           28.25  -0.214  -0.304   
1     Taipei     2000-01-03         4252.0           22.20   0.054  -0.065   
2     Taipei     2000-01-04         5486.0           23.40  -0.160  -0.152   
3     Taipei     2000-01-05         4841.0           19.65   0.056   0.056   
4     Taipei     2000-01-06         5049.0           20.75  -0.043  -0.183   
...      ...            ...            ...             ...     ...     ...   
6414  Taipei     2021-03-26        18932.0           32.95   0.042  -0.231   
6415  Taipei     2021-03-27        23591.0           34.35  -0.118  -0.360   
6416  Taipei     2021-03-28        31770.0           30.30  -0.163     nan   
6417  Taipei     2021-03-30        61186.0           25.35  -0.132     nan   
6418  Taipei     2021-03-31        39858.0           22.00     nan     nan   

         T+5     T+7  
0     -0.297  -0.400  
1     -0.065  -0.106  
2     -0.276  -0.113  
3      0.010   0.354  
4      0.000   0.200  
...      ...     ...  
6414     nan     nan  
6415     nan     nan  
6416     nan     nan  
6417     nan     nan  
6418     nan     nan  

[6419 rows x 8 columns]

In [35]:
df8['T+1']=df8['T+1'].astype(float)
df8['T+3']=df8['T+3'].astype(float)
df8['T+5']=df8['T+5'].astype(float)
df8['T+7']=df8['T+7'].astype(float)

In [36]:
df8.dtypes

Market                    object
ObservDay_time    datetime64[ns]
TradingVolume            float64
MediumAvgPrice           float64
T+1                      float64
T+3                      float64
T+5                      float64
T+7                      float64
dtype: object

In [37]:
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice    T+1    T+3  \
0     Taipei     2000-01-02         7348.0           28.25 -0.214 -0.304   
1     Taipei     2000-01-03         4252.0           22.20  0.054 -0.065   
2     Taipei     2000-01-04         5486.0           23.40 -0.160 -0.152   
3     Taipei     2000-01-05         4841.0           19.65  0.056  0.056   
4     Taipei     2000-01-06         5049.0           20.75 -0.043 -0.183   
...      ...            ...            ...             ...    ...    ...   
6414  Taipei     2021-03-26        18932.0           32.95  0.042 -0.231   
6415  Taipei     2021-03-27        23591.0           34.35 -0.118 -0.360   
6416  Taipei     2021-03-28        31770.0           30.30 -0.163    NaN   
6417  Taipei     2021-03-30        61186.0           25.35 -0.132    NaN   
6418  Taipei     2021-03-31        39858.0           22.00    NaN    NaN   

        T+5    T+7  
0    -0.297 -0.400  
1    -0.065 -0.106  
2    -0.276 -0.113  
3     0.010  0.354  
4     0.000  0.200  
...     ...    ...  
6414    NaN    NaN  
6415    NaN    NaN  
6416    NaN    NaN  
6417    NaN    NaN  
6418    NaN    NaN  

[6419 rows x 8 columns]

In [133]:
# MediumAvgPrice=df8.pop('MediumAvgPrice')
# df8.insert(2,'MediumAvgPrice',MediumAvgPrice)

In [38]:
df8

Market ObservDay_time  TradingVolume  MediumAvgPrice    T+1    T+3  \
0     Taipei     2000-01-02         7348.0           28.25 -0.214 -0.304   
1     Taipei     2000-01-03         4252.0           22.20  0.054 -0.065   
2     Taipei     2000-01-04         5486.0           23.40 -0.160 -0.152   
3     Taipei     2000-01-05         4841.0           19.65  0.056  0.056   
4     Taipei     2000-01-06         5049.0           20.75 -0.043 -0.183   
...      ...            ...            ...             ...    ...    ...   
6414  Taipei     2021-03-26        18932.0           32.95  0.042 -0.231   
6415  Taipei     2021-03-27        23591.0           34.35 -0.118 -0.360   
6416  Taipei     2021-03-28        31770.0           30.30 -0.163    NaN   
6417  Taipei     2021-03-30        61186.0           25.35 -0.132    NaN   
6418  Taipei     2021-03-31        39858.0           22.00    NaN    NaN   

        T+5    T+7  
0    -0.297 -0.400  
1    -0.065 -0.106  
2    -0.276 -0.113  
3     0.010  0.354  
4     0.000  0.200  
...     ...    ...  
6414    NaN    NaN  
6415    NaN    NaN  
6416    NaN    NaN  
6417    NaN    NaN  
6418    NaN    NaN  

[6419 rows x 8 columns]

In [39]:
df8.isnull().sum()

Market            0
ObservDay_time    0
TradingVolume     0
MediumAvgPrice    0
T+1               1
T+3               3
T+5               5
T+7               7
dtype: int64

In [40]:
df8.to_csv('/Users/lunlun/Desktop/BDSE_18_Project/data/蔬菜/FF1_veg.csv',index=False)

### test

In [1]:
import pandas as pd
df = pd.read_csv('/Users/lunlun/Desktop/BDSE_18_Project/data/蔬菜/LF2_veg.csv')

In [2]:
df

Market ObservDay_time  MediumAvgPrice  TradingVolume    T+1    T+3  \
0     Taipei     2000-01-02           54.40         3094.0  0.285  0.021   
1     Taipei     2000-01-03           69.90         2309.0 -0.006 -0.281   
2     Taipei     2000-01-04           69.45         2113.0 -0.200 -0.215   
3     Taipei     2000-01-05           55.55         1765.0 -0.095  0.012   
4     Taipei     2000-01-06           50.25         1317.0  0.085  0.047   
...      ...            ...             ...            ...    ...    ...   
6414  Taipei     2021-03-26           33.65         9660.0  0.049  0.171   
6415  Taipei     2021-03-27           35.30        11035.0  0.135 -0.011   
6416  Taipei     2021-03-28           40.05        10351.0 -0.016    NaN   
6417  Taipei     2021-03-30           39.40        13285.0 -0.114    NaN   
6418  Taipei     2021-03-31           34.90        11873.0    NaN    NaN   

        T+5    T+7  
0     0.002 -0.033  
1    -0.196 -0.326  
2    -0.243 -0.505  
3    -0.152 -0.453  
4    -0.316 -0.266  
...     ...    ...  
6414    NaN    NaN  
6415    NaN    NaN  
6416    NaN    NaN  
6417    NaN    NaN  
6418    NaN    NaN  

[6419 rows x 8 columns]

In [4]:
df.isnull().sum()

Market            0
ObservDay_time    0
MediumAvgPrice    0
TradingVolume     0
T+1               1
T+3               3
T+5               5
T+7               7
dtype: int64